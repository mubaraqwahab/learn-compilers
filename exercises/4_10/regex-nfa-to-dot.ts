import { writeFile, mkdir } from "node:fs/promises";
import { regexToNFA } from "./regex-to-nfa.js";

/**
 * Return the NFA corresponding to the given regex
 * in Graphviz DOT language format.
 */
export function regexNFAToDot(regex: string): string {
  const nfa = regexToNFA(regex);
  let dotText = "";

  nfa.transitions.forEach(([from, char, to]) => {
    dotText += `  ${from} -> ${to} [label=${labelify(char)}];\n`;
  });

  dotText = `digraph {\n  rankdir=LR;\n${dotText}}`;

  return dotText;
}

/**
 * Dump the DOT language representation of a regex in a given file.
 */
export async function dumpRegexNFA(regex: string, filePath: string) {
  const dotText = regexNFAToDot(regex);
  await writeFile(filePath, dotText);
}

function labelify(char: string): string {
  return JSON.stringify(char || "&epsilon;");
}

// (async () => {
//   await mkdir("graphs", { recursive: true });
//   dumpRegexNFA("a", "graphs/simple.dot");
//   dumpRegexNFA("ab", "graphs/concat.dot");
//   dumpRegexNFA("a|b", "graphs/union.dot");
//   dumpRegexNFA("a*", "graphs/kleene.dot");
//   dumpRegexNFA("(ab|c)*abb", "graphs/complex.dot");
// })();
