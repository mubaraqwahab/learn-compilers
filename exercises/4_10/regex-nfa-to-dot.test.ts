import { expect, test } from "vitest";
import { regexNFAToDot } from "./regex-nfa-to-dot.js";

expect.addSnapshotSerializer({
  test(val) {
    return typeof val === "string" && val.startsWith("digraph");
  },
  serialize(val) {
    // Return the DOT string as is so that I can copy
    // the DOT lang output from the snapshot file and
    // use it where I want without modifying it further.
    return val;
  },
});

test("can convert a simple regex to DOT lang", () => {
  expect(regexNFAToDot(`a`)).toMatchSnapshot();
});

test("can convert a concat to DOT lang", () => {
  expect(regexNFAToDot(`ab`)).toMatchSnapshot();
});

test("can convert a union to DOT lang", () => {
  expect(regexNFAToDot(`a|b`)).toMatchSnapshot();
});

test("can convert a kleene closure to DOT lang", () => {
  expect(regexNFAToDot(`a*`)).toMatchSnapshot();
});

test("can convert a complex regex to DOT lang", () => {
  expect(regexNFAToDot(`(ab|c)*abb`)).toMatchSnapshot();
});
