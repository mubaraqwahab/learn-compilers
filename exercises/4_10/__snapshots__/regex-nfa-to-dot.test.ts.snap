// Vitest Snapshot v1

exports[`can convert a complex regex to DOT lang 1`] = `
digraph {
  rankdir=LR;
  s0 -> s1 [label="&epsilon;"];
  s1 -> s2 [label="&epsilon;"];
  s2 -> s3 [label="a"];
  s3 -> s4 [label="b"];
  s1 -> s5 [label="&epsilon;"];
  s5 -> s6 [label="c"];
  s4 -> s7 [label="&epsilon;"];
  s6 -> s7 [label="&epsilon;"];
  s7 -> s1 [label="&epsilon;"];
  s7 -> s8 [label="&epsilon;"];
  s0 -> s8 [label="&epsilon;"];
  s8 -> s9 [label="a"];
  s9 -> s10 [label="b"];
  s10 -> s11 [label="b"];
}
`;

exports[`can convert a concat to DOT lang 1`] = `
digraph {
  rankdir=LR;
  s0 -> s1 [label="a"];
  s1 -> s2 [label="b"];
}
`;

exports[`can convert a kleene closure to DOT lang 1`] = `
digraph {
  rankdir=LR;
  s0 -> s1 [label="&epsilon;"];
  s1 -> s2 [label="a"];
  s2 -> s1 [label="&epsilon;"];
  s2 -> s3 [label="&epsilon;"];
  s0 -> s3 [label="&epsilon;"];
}
`;

exports[`can convert a simple regex to DOT lang 1`] = `
digraph {
  rankdir=LR;
  s0 -> s1 [label="a"];
}
`;

exports[`can convert a union to DOT lang 1`] = `
digraph {
  rankdir=LR;
  s0 -> s1 [label="&epsilon;"];
  s1 -> s2 [label="a"];
  s0 -> s3 [label="&epsilon;"];
  s3 -> s4 [label="b"];
  s2 -> s5 [label="&epsilon;"];
  s4 -> s5 [label="&epsilon;"];
}
`;
